sql/sql_base.cc

```diff
@@ -3846,6 +3846,45 @@ open_and_process_table(THD *thd, TABLE_LIST *tables, uint *counter, uint flags,
 
     thd->pop_internal_handler();
     safe_to_ignore_table= no_such_table_handler.safely_trapped_errors();
+
+#ifdef WITH_WSREP
+    /*
+      Append a table level shared key for the referenced/foreign table for:
+        - statement that updates existing rows (UPDATE, multi-update)
+        - statement that deletes existing rows (DELETE, DELETE_MULTI)
+      This is done to avoid potential MDL conflicts with concurrent DDLs.
+
+      The table level shared key are not appended if the wsrep client state is
+      not s_exec, as in the case during statement prepare it perform semantic
+      analysis of the parsed tree and send a response packet to the client.
+      While doing sematic analysis it opens all required tables to check access
+      rights. These operations to test prepared statements does not require to
+      append table level shared key, they will be added later.
+    */
+    if (tables->prelocking_placeholder == TABLE_LIST::PRELOCK_FK &&
+        WSREP(thd) && !thd->wsrep_applier && !error &&
+        wsrep_is_active(thd) &&
+        thd->wsrep_cs().state() == wsrep::client_state::s_exec &&
+        (sql_command_flags[lex->sql_command] &
+         (CF_UPDATES_DATA | CF_DELETES_DATA)))
```

Comment:
You might want to use trg_event_map or slave_fk_event_map in TABLE_LIST. Take a look at `DML_prelocking_strategy::handle_table` and `prepare_fk_prelocking_list` - that might be a more apropriate place for this code.

Reply:
I will be extra happy if you will find the way to contain all this designation and handling code inside some wsrep*.cc file. Anyways, it's reasonable to extract it into a separate function.